atomic_reactor/pulp_util.py

```diff
@@ -224,3 +227,55 @@ def get_registry_hostname(self):
 
     def get_pulp_instance(self):
         return self.pulp_instance
+
+
+class LockedPulpRepository(object):
+    """
+    Context manager for Pulp repository operations.
+
+    This context manager creates a 'lock' repository on entry, and
+    deletes the repository on exit.
+
+    Usage:
+
+    with LockedPulpRepository(handler, repo_id) as locked_repo:
+        locked_repo.set_image_tags(v1_image_id, tags)
+    """
+
+    def __init__(self, pulp, repo_id, prefix='lock-'):
+        self.pulp = pulp
+        self.repo_id = repo_id
+        self.prefix = prefix
+
+        self.retry_times = 9
+        self.retry_delay = 5
+
+    def __enter__(self):
+        logger.info("creating lock repository %s%s", self.prefix, self.repo_id)
+        registry_id = 'lock/{}'.format(self.repo_id)
+        retry = 0
+        for counter in range(self.retry_times + 1):
+            if retry:
+                time.sleep(retry)
+
+            try:
+                self.pulp.createRepo(self.repo_id, None,
+                                     registry_id=registry_id,
+                                     is_origin=True,  # avoid origin-lock-...
+                                     prefix_with=self.prefix)
+                break
+            except dockpulp.DockPulpError as exc:
+                if counter == self.retry_times:
+                    # This was the last chance; go ahead anyway. This
+                    # avoids a stale lock preventing builds.
+                    logging.info("got error, breaking lock: %s", exc)
```

Comment:
@lcarva, not sure if you spotted this, but I've made it more explicit that we will break the lock if we have to.

Reply:
I think that's fine. The update repo operation is pretty fast, unlike things like publish.